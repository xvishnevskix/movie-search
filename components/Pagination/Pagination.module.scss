.pagination {
  position: relative;
  --border: 1px solid #d3d3d3;
  height: 50px;
  display: inline-flex;
  align-items: center;
  border-radius: 5px;
  overflow: hidden;
  justify-content: space-between;

  &:before {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border-radius: 5px;
    overflow: hidden;
    border: var(--border);
  }
}

.btn {
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  width: 50px;
  height: 100%;
  color: var(--color-dark);
  transition: background-color 0.3s, color 0.3s, border 0.3s;

  &[disabled] {
    opacity: 0.5;
    pointer-events: none;
  }

  &:hover {
    color: var(--color-white);
    border-color: transparent;
    background-color: var(--color-primary);
  }
}

.left {
  border-right: var(--border);
}

.right {
  border-left: var(--border);
}

.value {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 20px;
  height: 100%;
  border-right: var(--border);
  border-left: var(--border);
}