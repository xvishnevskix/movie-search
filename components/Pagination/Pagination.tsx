import { FC, memo, useEffect } from 'react'
import { FiChevronLeft, FiChevronRight } from 'react-icons/fi';
import { useTypedSelector } from '@/hooks/useTypedSelector';
import { useActions } from '@/hooks/useActions';
import classNames from 'classnames';

interface PaginationProps {
  pages: number | any;
  classN?: string;
}

export const Pagination = memo<PaginationProps>(({pages, classN}) => {

  const {page} = useTypedSelector(state => state.paginationReducer)
  const {setPage} = useActions()
  const array = Array.from({length: pages}, (_,idx) => idx + 1)
  const handleBack = () => setPage(page - 1)
  const handleForward = () => setPage(page + 1)
  const handleSetPage = (page: number) => setPage(page)
  const isMaxPages = pages <= 5 ? pages : 5 

  if(array.length > 5) array.length = 5

  return (
    <>
      {pages !== 1 &&
        <ul className={classNames('list-reset g-pagination', classN)}>
          <li className='g-pagination__item'>
            <button
              onClick={handleBack}
              className={classNames('btn-reset g-pagination__btn g-pagination__link--prev', page === 1 && 'g-pagination__btn--disabled')}
            >
              <FiChevronLeft />
            </button>
          </li>
          {array?.map(el => (
            <li key={el} className='g-pagination__item'>
              {page === el
              ?
                <span className={classNames('g-pagination__btn g-pagination__btn--current')}>
                  {el}
                </span>
              :
                <button onClick={() => handleSetPage(el)} className='btn-reset g-pagination__btn'>
                  {el}
                </button>
              }
            </li>
          ))}
          <li className='g-pagination__item'>
            <button
              onClick={handleForward}
              className={classNames('btn-reset g-pagination__btn g-pagination__link--next', page === isMaxPages && 'g-pagination__btn--disabled')}
            >
              <FiChevronRight />
            </button>
          </li>
        </ul>
      }
    </>
  )
})

Pagination.displayName = 'Pagination'