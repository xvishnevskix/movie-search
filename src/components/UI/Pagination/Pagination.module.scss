.pagination {
	position: relative;
	--border: 1px solid #d3d3d3;
	--size: 50px;
	height: var(--size);
	display: flex;
	align-items: center;
	border-radius: 5px;
	overflow: hidden;
	width: fit-content;

	@media (max-width: 1024px) {
		margin: 0 auto;
	}

	@media (max-width: 576px) {
		--size: 43px;
	}

	&:before {
		content: '';
		position: absolute;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
		border-radius: 5px;
		overflow: hidden;
		border: var(--border);
	}
}

.btn {
	display: inline-flex;
	align-items: center;
	justify-content: center;
	position: relative;
	width: var(--size);
	height: 100%;
	color: var(--color-dark);
	transition: background-color 0.3s, color 0.3s, border 0.3s;

	&[disabled] {
		opacity: 0.5;
		pointer-events: none;
	}

	&:not([disabled]) {
		&:hover {
			color: var(--color-white);
			border-color: transparent;
			background-color: var(--color-primary);
		}
	}
}

.left {
	border-right: var(--border);
}

.right {
	border-left: var(--border);
}

.value {
	display: inline-flex;
	align-items: center;
	justify-content: center;
	padding: 0 20px;
	height: 100%;
	border-right: var(--border);
	border-left: var(--border);
}
