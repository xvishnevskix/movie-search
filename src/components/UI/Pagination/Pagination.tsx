import { FC, useEffect } from 'react';
import { FiChevronsLeft, FiChevronLeft, FiChevronsRight, FiChevronRight } from 'react-icons/fi';
import { ButtonBase } from '../ButtonBase/ButtonBase';
import styles from './Pagination.module.scss';
import classNames from 'classnames';

interface PaginationProps {
	setPage: (page: number) => void;
	pages: number | undefined;
	page: number;
	className?: string;
}

export const Pagination: FC<PaginationProps> = ({ page, setPage, pages, className }) => {
	const isFirstPage = page === 1;
	const isLastPage = page === pages;

	return (
		<>
			{pages !== 1 && (
				<div data-testid="pagination" className={classNames(styles.pagination, className)}>
					<ButtonBase
						data-testid="pagination-first-btn"
						onClick={() => setPage(1)}
						className={classNames(styles.btn, styles.left)}
						disabled={isFirstPage}
					>
						<FiChevronsLeft />
					</ButtonBase>
					<ButtonBase
						data-testid="pagination-prev-btn"
						onClick={() => setPage(page - 1)}
						className={styles.btn}
						disabled={isFirstPage}
					>
						<FiChevronLeft />
					</ButtonBase>
					<span data-testid="pagination-value" className={styles.value}>
						{page} / {pages}
					</span>
					<ButtonBase
						data-testid="pagination-next-btn"
						disabled={isLastPage}
						onClick={() => setPage(page + 1)}
						className={styles.btn}
					>
						<FiChevronRight />
					</ButtonBase>
					<ButtonBase
						data-testid="pagination-last-btn"
						disabled={isLastPage}
						onClick={() => setPage(Number(pages))}
						className={classNames(styles.btn, styles.right)}
					>
						<FiChevronsRight />
					</ButtonBase>
				</div>
			)}
		</>
	);
};
