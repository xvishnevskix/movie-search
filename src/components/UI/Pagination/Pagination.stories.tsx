import React, { useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Pagination } from './Pagination';

export default {
	title: 'Pagination',
	component: Pagination,
} as ComponentMeta<typeof Pagination>;

const Template: ComponentStory<typeof Pagination> = () => {
	const [page, setPage] = useState<number>(1);

	return <Pagination setPage={setPage} page={page} pages={100} />;
};

export const Default = Template.bind({});
