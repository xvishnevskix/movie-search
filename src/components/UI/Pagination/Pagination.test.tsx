import { Pagination } from './Pagination';
import { render, screen } from '@testing-library/react';
import { useState } from 'react';
import userEvent from '@testing-library/user-event';

describe('пагинация', () => {
	beforeEach(() => jest.useFakeTimers());
	window.scrollTo = jest.fn();

	const Wrapper = () => {
		const [page, setPage] = useState<number>(1);
		return <Pagination page={page} setPage={setPage} pages={100} />;
	};

	test('проверка всё ли ок', () => {
		render(<Wrapper />);
		const element = screen.getByTestId('pagination');
		expect(element).toBeInTheDocument();
		expect(element).toMatchSnapshot();
	});

	test('работает ли клик по кнопке вперёд', () => {
		const { asFragment } = render(<Wrapper />);

		userEvent.click(screen.getByTestId('pagination-next-btn'));
		jest.runAllTimers();
		expect(screen.getByTestId('pagination-value')).toHaveTextContent('2 / 100');

		expect(asFragment()).toMatchSnapshot();
	});

	test('работает ли клик по кнопке назад', () => {
		const { asFragment } = render(<Wrapper />);

		userEvent.click(screen.getByTestId('pagination-prev-btn'));
		jest.runAllTimers();
		expect(screen.getByTestId('pagination-value')).toHaveTextContent('1 / 100');

		expect(asFragment()).toMatchSnapshot();
	});
	test('работает ли клик по кнопке последняя страница', () => {
		const { asFragment } = render(<Wrapper />);

		userEvent.click(screen.getByTestId('pagination-last-btn'));
		jest.runAllTimers();
		expect(screen.getByTestId('pagination-value')).toHaveTextContent('100 / 100');

		expect(asFragment()).toMatchSnapshot();
	});

	test('работает ли клик по кнопке первая страница', () => {
		const { asFragment } = render(<Wrapper />);

		userEvent.click(screen.getByTestId('pagination-first-btn'));
		jest.runAllTimers();
		expect(screen.getByTestId('pagination-value')).toHaveTextContent('1 / 100');

		expect(asFragment()).toMatchSnapshot();
	});
});
