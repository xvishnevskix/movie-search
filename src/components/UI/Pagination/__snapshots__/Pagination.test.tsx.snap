// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`пагинация проверка всё ли ок 1`] = `
<div
  class="pagination"
  data-testid="pagination"
>
  <button
    class="btn btn left"
    data-testid="pagination-first-btn"
    disabled=""
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="11 17 6 12 11 7"
      />
      <polyline
        points="18 17 13 12 18 7"
      />
    </svg>
  </button>
  <button
    class="btn btn"
    data-testid="pagination-prev-btn"
    disabled=""
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="15 18 9 12 15 6"
      />
    </svg>
  </button>
  <span
    class="value"
    data-testid="pagination-value"
  >
    1
     / 
    100
  </span>
  <button
    class="btn btn"
    data-testid="pagination-next-btn"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="9 18 15 12 9 6"
      />
    </svg>
  </button>
  <button
    class="btn btn right"
    data-testid="pagination-last-btn"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="13 17 18 12 13 7"
      />
      <polyline
        points="6 17 11 12 6 7"
      />
    </svg>
  </button>
</div>
`;

exports[`пагинация работает ли клик по кнопке вперёд 1`] = `
<DocumentFragment>
  <div
    class="pagination"
    data-testid="pagination"
  >
    <button
      class="btn btn left"
      data-testid="pagination-first-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="11 17 6 12 11 7"
        />
        <polyline
          points="18 17 13 12 18 7"
        />
      </svg>
    </button>
    <button
      class="btn btn"
      data-testid="pagination-prev-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </button>
    <span
      class="value"
      data-testid="pagination-value"
    >
      2 / 100
    </span>
    <button
      class="btn btn"
      data-testid="pagination-next-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </button>
    <button
      class="btn btn right"
      data-testid="pagination-last-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="13 17 18 12 13 7"
        />
        <polyline
          points="6 17 11 12 6 7"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`пагинация работает ли клик по кнопке назад 1`] = `
<DocumentFragment>
  <div
    class="pagination"
    data-testid="pagination"
  >
    <button
      class="btn btn left"
      data-testid="pagination-first-btn"
      disabled=""
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="11 17 6 12 11 7"
        />
        <polyline
          points="18 17 13 12 18 7"
        />
      </svg>
    </button>
    <button
      class="btn btn"
      data-testid="pagination-prev-btn"
      disabled=""
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </button>
    <span
      class="value"
      data-testid="pagination-value"
    >
      1 / 100
    </span>
    <button
      class="btn btn"
      data-testid="pagination-next-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </button>
    <button
      class="btn btn right"
      data-testid="pagination-last-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="13 17 18 12 13 7"
        />
        <polyline
          points="6 17 11 12 6 7"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`пагинация работает ли клик по кнопке первая страница 1`] = `
<DocumentFragment>
  <div
    class="pagination"
    data-testid="pagination"
  >
    <button
      class="btn btn left"
      data-testid="pagination-first-btn"
      disabled=""
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="11 17 6 12 11 7"
        />
        <polyline
          points="18 17 13 12 18 7"
        />
      </svg>
    </button>
    <button
      class="btn btn"
      data-testid="pagination-prev-btn"
      disabled=""
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </button>
    <span
      class="value"
      data-testid="pagination-value"
    >
      1 / 100
    </span>
    <button
      class="btn btn"
      data-testid="pagination-next-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </button>
    <button
      class="btn btn right"
      data-testid="pagination-last-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="13 17 18 12 13 7"
        />
        <polyline
          points="6 17 11 12 6 7"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`пагинация работает ли клик по кнопке последняя страница 1`] = `
<DocumentFragment>
  <div
    class="pagination"
    data-testid="pagination"
  >
    <button
      class="btn btn left"
      data-testid="pagination-first-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="11 17 6 12 11 7"
        />
        <polyline
          points="18 17 13 12 18 7"
        />
      </svg>
    </button>
    <button
      class="btn btn"
      data-testid="pagination-prev-btn"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </button>
    <span
      class="value"
      data-testid="pagination-value"
    >
      100 / 100
    </span>
    <button
      class="btn btn"
      data-testid="pagination-next-btn"
      disabled=""
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </button>
    <button
      class="btn btn right"
      data-testid="pagination-last-btn"
      disabled=""
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="13 17 18 12 13 7"
        />
        <polyline
          points="6 17 11 12 6 7"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
